File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        if(root != null){
            this.helper(root);
        }
    }
    
    private TreeNode[] helper(TreeNode curr){
        TreeNode tail = curr, left = curr.left, right = curr.right;
        
        if(left != null){
            TreeNode[] res = this.helper(left);
            tail.right = res[0];
            tail = res[1];
        }
        
        curr.left = null;
        
        if(right != null){
            TreeNode[] res = this.helper(right);
            tail.right = res[0];
            tail = res[1];
        }
        
        return new TreeNode[]{curr, tail};
    }
}

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return None
        
        def helper(curr: Optional[TreeNode]) -> Tuple[TreeNode, TreeNode]:
            tail, left, right = curr, curr.left, curr.right
            
            if left:
                tail.right, tail = helper(left)
            
            curr.left = None
                
            if right:
                tail.right, tail = helper(right)
            
            return curr, tail
        
        helper(root)